pkg/ruby/check_upgrade.go

```diff
@@ -0,0 +1,199 @@
+	cachedPath := o.cachedGemspecPath(pkg, file)
+	cached, err := os.Open(cachedPath)
+	if err != nil || o.NoCache {
+		ctx := context.Background()
+
+		client := github.NewClient(o.Client.Client)
+		gitOpts := gh.GitOptions{
+			GithubClient: client,
+			Logger:       logger,
+		}
+
+		fileContent, err := gitOpts.RepositoryFilesContents(ctx, pkg.Repo.Organisation, pkg.Repo.Name, file, pkg.Ref)
+		if err != nil {
+			return "", err
+		}
+
+		// Decode the base64-encoded content
+		decodedContent, err := base64.StdEncoding.DecodeString(*fileContent.Content)
+		if err != nil {
+			return "", fmt.Errorf("error decoding file content: %w", err)
+		}
+
+		err = os.MkdirAll(path.Dir(cachedPath), 0o755)
+		if err != nil {
+			return "", fmt.Errorf("failed to create cache directory: %w", err)
+		}
+
+		cached, err = os.Create(cachedPath)
+		if err != nil {
+			return "", fmt.Errorf("failed to create cache file: %w", err)
+		}
+
+		_, err = cached.Write(decodedContent)
+		if err != nil {
+			return "", fmt.Errorf("failed to write cache file: %w", err)
+		}
+	}
+	defer cached.Close()
+	return cachedPath, nil
+}
+
+// checkVersionConstraint takes a gemspec and extracts the version constraints
+// and checks them against the RubyUpdateVersion. This function attempts to use
+// inline ruby to determine if the version is within constraints. If ruby is
+// not installed on the system then the terraform go-version package is used as
+// it has very similar version constraints. The inline ruby is the preferred
+// method in case something changes (highly unlikely).
+//
+// NOTE: If there is no required_ruby_version found in the gemspec then we
+//
+//	assume there is no limit.
+func (o *Options) checkVersionConstraint(gemspecFile string) error {
+	cached, err := os.Open(gemspecFile)
+	if err != nil {
+		return fmt.Errorf("opening gemspec file: %w", err)
+	}
+	versionLine := ""
+	scanner := bufio.NewScanner(cached)
+	for scanner.Scan() {
+		if strings.Contains(scanner.Text(), requiredRubyVersionKey) {
+			versionLine = scanner.Text()
+			break
+		}
+	}
+
+	// Could not find required_ruby_version in gemspec, assuming no constraint
+	if versionLine == "" {
+		return nil
+	}
+
+	pattern := `["']([^"']+)["']`
+	re := regexp.MustCompile(pattern)
+	groups := re.FindAllStringSubmatch(versionLine, -1)
+
+	versionConstraints := []string{}
+	for _, group := range groups {
+		versionConstraints = append(versionConstraints, group[1])
+	}
+
+	if len(versionConstraints) < 1 {
+		return fmt.Errorf("could not extract %s from gemspec", requiredRubyVersionKey)
+	}
+
+	rubyPath, err := exec.LookPath("ruby")
+	if err == nil { // Call ruby to do the comparison
+		// Need to construct a list of version constraints to pass
+		// to the ruby call. It should end up like this: ['>= 2.6', '< 4']
+		constraint := "["
+		for i, c := range versionConstraints {
+			constraint += fmt.Sprintf("'%s'", c)
+			if i != len(versionConstraints)-1 {
+				constraint += ", "
+			}
+		}
+		constraint += "]"
```

Comment:
Can this be simplified with [`strings.Join()`](https://pkg.go.dev/strings#Join)?
I'm thinking something like:
```go
constraint := fmt.Sprintf("['%s']", strings.Join(versionConstraints, "', '"))
```

Reply:
Love it!